Intersect.Client/Entities/Player.cs

```diff
@@ -799,46 +840,79 @@ public void TryWithdrawItem(int bankSlotIndex, int inventorySlotIndex = -1)
 
-            if (itemDescriptor.IsStackable)
+            if (availableSpace < 1)
             {
-                maxQuantity = GetQuantityOfItemInBank(itemDescriptor.Id);
+                ChatboxMsg.AddMessage(new ChatboxMsg(Strings.Bank.WithdrawItemNoSpace, CustomColors.Alerts.Error, ChatMessageType.Bank));
+                return;
             }
 
-            if (maxQuantity < 2)
+            if (itemQuantityInBank < 2)
             {
-                PacketSender.SendWithdrawItem(bankSlotIndex, 1, inventorySlotIndex);
+                WithdrawSingleItem(bankSlotIndex, inventorySlotIndex);
                 return;
             }
 
-            int[] userData = new int[2] { bankSlotIndex, inventorySlotIndex };
+            OpenWithdrawItemInputBox(itemDescriptor, bankSlotIndex, inventorySlotIndex, availableSpace);
+        }
 
+        private bool IsGuildBankWithdrawAllowed()
+        {
+            var rank = Globals.Me.GuildRank;
+            return !string.IsNullOrWhiteSpace(Globals.Me.Guild) && (rank.Permissions.BankRetrieve || Globals.Me.Rank == 0);
+        }
+
+        private void WithdrawSingleItem(int bankSlotIndex, int inventorySlotIndex)
+        {
+            PacketSender.SendWithdrawItem(bankSlotIndex, 1, inventorySlotIndex);
+        }
+
+        private void OpenWithdrawItemInputBox(ItemBase itemDescriptor, int bankSlotIndex, int inventorySlotIndex, int availableSpace)
+        {
             InputBox.Open(
                 title: Strings.Bank.withdrawitem,
                 prompt: Strings.Bank.withdrawitemprompt.ToString(itemDescriptor.Name),
                 modal: true,
                 inputType: InputBox.InputType.NumericSliderInput,
                 onSuccess: WithdrawItemInputBoxOkay,
                 onCancel: null,
-                userData: userData,
-                quantity: quantity,
-                maxQuantity: maxQuantity
+                userData: new[] { bankSlotIndex, inventorySlotIndex },
+                quantity: availableSpace,
+                maxQuantity: availableSpace
             );
         }
 
+        private int FindAvailableInventorySpaceForItem(Guid itemId)
+        {
+            int spaceLeft = 0;
+            int maxInventoryStack = ItemBase.Get(itemId).MaxInventoryStack;
+
+            // Calculate the total available space in the player's inventory for this item.
+            for (int i = 0; i < Inventory.Length; i++)
+            {
+                var inventorySlot = Inventory[i];
+                if (inventorySlot != null && inventorySlot.ItemId == itemId)
+                {
+                    spaceLeft += maxInventoryStack - inventorySlot.Quantity;
+                }
+                else if (inventorySlot == null || inventorySlot.ItemId == Guid.Empty)
+                {
+                    spaceLeft += maxInventoryStack;
+                }
+            }
+
+            return spaceLeft;
+        }
```

Comment:
All of my comments re: scanning bank slots apply here.

1. Null-safety on `ItemBase.Get` (use `TryGet`)
2. `ToArray()` `Inventory` into a local variable to protect against race conditions
3. `foreach` over that local array
4. return as soon as you know that it exceeds the number that are even available in the bank (I do realize you don't even have a parameter for this yet but you said you are adding this in a bit)